drivers/usb/udc/udc_common.c

```diff
@@ -0,0 +1,1112 @@
+
+	buf = udc_ctrl_alloc_status(dev, dout, USB_CONTROL_EP_IN);
+	if (buf == NULL) {
+		ret = -ENOMEM;
+	}
+
+	return udc_submit_event(dev, UDC_EVT_EP_REQUEST, ret, data->setup);
+}
+
+int udc_ctrl_submit_s_in_status(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+	struct net_buf *buf;
+	int ret = 0;
+
+	if (!udc_ctrl_stage_is_data_in(dev)) {
+		return -ENOTSUP;
+	}
+
+	/* Allocate buffer for data stage IN */
+	buf = udc_ctrl_alloc_data(dev, data->setup, USB_CONTROL_EP_IN);
+	if (buf == NULL) {
+		ret = -ENOMEM;
+	}
+
+	return udc_submit_event(dev, UDC_EVT_EP_REQUEST, ret, data->setup);
+}
+
+int udc_ctrl_submit_s_status(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+	struct net_buf *buf;
+	int ret = 0;
+
+	/* Allocate buffer for possible status IN */
+	buf = udc_ctrl_alloc_status(dev, data->setup, USB_CONTROL_EP_IN);
+	if (buf == NULL) {
+		ret = -ENOMEM;
+	}
+
+	return udc_submit_event(dev, UDC_EVT_EP_REQUEST, ret, data->setup);
+}
+
+int udc_ctrl_submit_status(const struct device *dev,
+			   struct net_buf *const buf)
+{
+	struct udc_buf_info *bi = udc_get_buf_info(buf);
+
+	bi->status = true;
+
+	return udc_submit_event(dev, UDC_EVT_EP_REQUEST, 0, buf);
+}
+
+bool udc_ctrl_stage_is_data_out(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_DATA_OUT ? true : false;
+}
+
+bool udc_ctrl_stage_is_data_in(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_DATA_IN ? true : false;
+}
+
+bool udc_ctrl_stage_is_status_out(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_STATUS_OUT ? true : false;
+}
+
+bool udc_ctrl_stage_is_status_in(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_STATUS_IN ? true : false;
+}
+
+bool udc_ctrl_stage_is_no_data(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	return data->stage == CTRL_PIPE_STAGE_NO_DATA ? true : false;
+}
+
+void udc_ctrl_update_stage(const struct device *dev,
+			   struct net_buf *const buf)
+{
+	struct udc_buf_info *bi = udc_get_buf_info(buf);
+	struct udc_device_caps caps = udc_caps(dev);
+	uint8_t next_stage = CTRL_PIPE_STAGE_ERROR;
+	struct udc_data *data = dev->data;
+
+	__ASSERT(USB_EP_GET_IDX(bi->ep) == 0,
+		 "0x%02x is not a control endpoint", bi->ep);
+
+	if (bi->setup && bi->ep == USB_CONTROL_EP_OUT) {
+		uint16_t length  = udc_data_stage_length(buf);
+
+		data->setup = buf;
+
+		if (data->stage != CTRL_PIPE_STAGE_SETUP) {
+			LOG_INF("Sequence %u not completed", data->stage);
+			data->stage = CTRL_PIPE_STAGE_SETUP;
+		}
+
+		/*
+		 * Setup Stage has been completed (setup packet received),
+		 * regardless of the previous stage, this is now being reset.
+		 * Next state depends on the direction and
+		 * the value of wLength.
+		 */
+		if (length && udc_data_stage_to_device(buf)) {
```

Comment:
Universal Serial Bus Specification Revision 2.0 9.3.1
> The state of the Direction bit is ignored if the wLength field is zero, signifying there is no Data stage.

The simplest way to comply to the specification is to change the if chain to start with `if (length == 0)` then `else if (udc_data_stage_to_device(buf))` and finally `else` (data stage from device).